### PR TITLE
use indexmap for pdf_dict

### PR DESCRIPTION
**Pull request by burrbull**
Part of #176 with patches

### CODE REVIEW

**Comment by burrbull** [comment]
77 changes
And 40 minutes

File: dpx/src/dpx_pdfobj.rs [hunk 1]
```diff
@@ -231,12 +233,12 @@ impl Default for xref_entry {
     }
 }
 
-#[derive(Copy, Clone)]
+use indexmap::IndexMap;
+
+#[derive(Clone)]
```
Comment (mulimoen) [inline dpx/src/dpx_pdfobj.rs]: This clone does not deeply clone it's elements, and is very prone to use-after-free for its elements

File: dpx/src/dpx_pdfobj.rs [hunk 2]
```diff
@@ -1265,53 +1271,70 @@ impl pdf_dict {
...
-         * the new key just before the end
-         */
-        let new_node = Box::new(pdf_dict::new());
-        data.next = Box::into_raw(new_node);
-        data.key = name;
-        data.value = value;
-        0i32
     }
     /* pdf_merge_dict makes a link for each item in dict2 before stealing it */
     pub unsafe fn merge(&mut self, dict2: &Self) {
```
Comment (mulimoen) [inline dpx/src/dpx_pdfobj.rs:1279]: Does this mean the ownership of dict2 is transferred to &self? In which case we can take dict2 as Self

Comment (cormacrelf) [reply dpx/src/dpx_pdfobj.rs:1279]: I think it just increments a reference count on all of dict2’s values. Taking as Self would just mean some code has to pre-emptively clone dict2. Not advantageous.

**Comment by burrbull** [comment]
Looks like changes always in the end of files.

File: dpx/src/dpx_pdfobj.rs [hunk 3]
```diff
@@ -1281,14 +1280,10 @@ impl pdf_dict {
...
         for (k, &v) in dict2.inner.iter() {
-            let linked = pdf_link_obj(v);
-            self.inner
-                .entry(k.clone())
-                .and_modify(|existing| {
-                    let existing = mem::replace(existing, linked);
-                    pdf_release_obj(existing);
```
Comment (cormacrelf) [inline dpx/src/dpx_pdfobj.rs]: Releasing is important, right?

**Comment by cormacrelf** [comment]
Worth noting that these are exactly the same failures as #176 has overall

File: dpx/src/dpx_pdfobj.rs (hunk 3 above)
Comment (burrbull) [reply dpx/src/dpx_pdfobj.rs]: I think no, not here.

**Comment by burrbull** [comment]
> Worth noting that these are exactly the same failures as #176 has overall

I've noticed this.

File: dpx/src/dpx_pdfobj.rs (hunk 3 above)
Comment (cormacrelf) [reply dpx/src/dpx_pdfobj.rs]: 1. [The old code does it](https://github.com/crlf0710/tectonic/pull/212/files#diff-783cb601464af70fc34a0609b6535172L1272)
2. Indexmap does no Drop with raw pointers, so it is still necessary to match the old code
3. The codebase treats dicts as owning the values and doesn't free them after adding them to the dict
4. I don't see anything special about merging that would not require releasing the previous object when deleting and draining do require it

Comment (cormacrelf) [reply dpx/src/dpx_pdfobj.rs]: On the other hand, using `Rc` will steamroll this eventually.

**Comment by burrbull** [comment]
Rebased

File: dpx/src/dpx_pdfobj.rs (hunk 3 above)
Comment (burrbull) [reply dpx/src/dpx_pdfobj.rs]: `merge` use `pdf_dict`s `set` which release old value

**Comment by burrbull** [comment]
I've just added `dbg` in `pdf_write_obj`:

изображение

You can see result:

Снимок экрана_2019-11-24_08-49-57

**Comment by burrbull** [comment]
In 1702.00195 one change in dict write order

изображение


изображение

**Comment by burrbull** [comment]
I think difference in `pdf_remove_dict`.

When indexmap remove key, sorting changed:
```

Creator
Title
Subject
Author
Producer
Keywords

Removed:   Title

Creator
Keywords
Subject
Author
Producer
```

**Comment by burrbull** [comment]
Close in favour of #220

**Comment by mulimoen** [comment]
You can try using shift_remove on the indexmap. Default behaviour is swap_remove

**Comment by burrbull** [comment]
> You can try using shift_remove on the indexmap. Default behaviour is swap_remove

See #220

**Comment by cormacrelf** [comment]
Great job!

**Pull request closed without merging**